gpii/node_modules/singleInstance/src/singleInstance.js

```diff
@@ -0,0 +1,118 @@
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+"use strict";
+
+var fluid = require("universal");
+var fs = require("fs");
+
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.singleInstance");
+
+/**
+ * Declares this instance of GPII to be the only instance, unless there's already one running.
+ *
+ * The "pid/lock file" method is used, where a file containing the pid of the current process is created. If it
+ * already exists, and the pid points to a running GPII instance, then the lock will fail.
+ *
+ * File IO is performed synchronously, ensuring the checks are made before GPII does anything else (eg, bind to a port).
+ *
+ * @todo More checks need to be made to ensure the pid is GPII, and not just another process.
+ *
+ * @param pidFile {String} [Optional] The path of the pid file. Default is "gpii.pid" in the settings directory.
+ * @return {boolean} true if this instance is the only GPII instance.
+ */
+gpii.singleInstance.registerInstance = function (pidFile) {
+    if (!pidFile) {
+        var settingsDirComponent = gpii.settingsDir();
+        pidFile = settingsDirComponent.getGpiiSettingsDir() + "/gpii.pid";
+    }
+
+    var gpiiPid = gpii.singleInstance.checkInstance(pidFile);
+    if (!gpiiPid) {
+        // Write the pid, but fail if the file exists. The stale one was deleted, but there might be another
+        // GPII instance just starting.
+        fs.writeFileSync(pidFile, process.pid, { flag: "wx" });
+    }
+
+    return !gpiiPid || gpiiPid === process.pid;
+};
+
+/**
+ * Checks for another GPII instance.
+ *
+ * @param pidFile {String} [Optional] The path of the pid file. Default is "gpii.pid" in the settings directory.
+ * @returns {number} the PID of the other GPII process (including this process), or null if this is the only one.
+ */
+gpii.singleInstance.checkInstance = function (pidFile) {
+    if (!pidFile) {
+        var settingsDirComponent = gpii.settingsDir();
+        pidFile = settingsDirComponent.getGpiiSettingsDir() + "/gpii.pid";
+    }
+
+    var pid = null;
+
+    try {
+        // Get the old pid from the lock file
+        var content = fs.readFileSync(pidFile, {encoding: "utf8"});
+        pid = parseInt(content);
+        if (pid && pid !== process.pid) {
+            try {
+                // Check if it's still running.
+                process.kill(pid, 0);
+                // No error means the process is running.
+                // TODO: Ensure the process really is GPII, and not just a re-use of the pid.
+            } catch (e) {
+                // Process isn't running (or permission denied, which suggests it's not GPII).
+                pid = null;
+            }
+        }
+
+        if (!pid) {
+            // Remove the pid file, as it's stale/invalid.
+            fs.unlinkSync(pidFile);
+        }
+    } catch (e) {
+        // The file doesn't exist.
+        pid = null;
```

Comment:
I believe there are no code paths whereby an exception could reach here with pid being set that aren't caught by the inner try block. Unlike C++, JavaScript doesn't permit "exceptions on assignment" : P